runtime/major_gc.c

```diff
@@ -724,29 +802,102 @@ static void mark_slice_darken(struct mark_stack* stk, value child,
         }
       }
     }
   }
+  return 0;
+}
+
+static void mark_slice_darken(struct mark_stack* stk, value child,
+                              intnat* work)
+{
+  child = mark_slice_darken_only(child, work);
+  if (child) {
+    *work -= mark_stack_push_block(stk, child);
+  }
 }
 
-static intnat do_some_marking(struct mark_stack* stk, intnat budget) {
-  while (stk->count > 0) {
-    mark_entry me = stk->stack[--stk->count];
-    while (me.start < me.end) {
-      if (budget <= 0) {
-        mark_stack_push_range(stk, me.start, me.end);
-        return budget;
+Caml_noinline static intnat do_some_marking(struct mark_stack* stk,
+                                            intnat budget) {
+  prefetch_buffer_t pb = { .enqueued = 0, .dequeued = 0,
+                           .waterline = PREFETCH_BUFFER_MIN };
+  mark_entry me;
+
+  while (1) {
+    if (pb_above_waterline(&pb)) {
+      /* Dequeue from prefetch buffer */
+      value block = pb_pop(&pb);
+      CAMLassert(Is_markable(block));
+
+      block = mark_slice_darken_only(block, &budget);
+      budget--; /* header word */
+      if (!block)
+        continue;
+
+      const header_t hd = Hd_val(block);
+      me.start = Op_val(block);
+      me.end = me.start + Wosize_hd(hd);
+
+      if (Tag_hd(hd) == Closure_tag) {
+        uintnat env_offset = Start_env_closinfo(Closinfo_val(block));
+        budget -= env_offset;
+        me.start += env_offset;
+      }
+    }
+    else if (budget <= 0 || stk->count == 0) {
+      if (pb.waterline > 0) {
+        /* Dequeue from pb even when close to empty, because
+           we have nothing else to do */
+        pb_drain(&pb);
+        continue;
+      }
+      else {
+        /* Couldn't find work with pb in draining mode,
+           so there's nothing to do */
+        break;
+      }
+    }
+    else {
+      me = stk->stack[--stk->count];
+    }
+
+    value* scan_end = me.end;
+    if (scan_end - me.start > budget) {
+      intnat scan_len = budget < 0 ? 0 : budget;
+      scan_end = me.start + scan_len;
+    }
+
+    for (; me.start < scan_end; me.start++) {
+      CAMLassert(budget >= 0);
+
+      value child = *me.start;
+      if (Is_markable(child)) {
+        if (pb_full(&pb))
+          break;
+        prefetch_block(child);
+        pb_push(&pb, child);
+      }
+    }
+
+    if (me.start < me.end) {
+      /* Didn't finish scanning this object, either because budget <= 0,
+         or the prefetch buffer filled up. Leave the rest on the stack. */
+      mark_stack_push_range(stk, me.start, me.end);
+      caml_prefetch(me.start+1);
+
+      if (pb_size(&pb) > PREFETCH_BUFFER_MIN) {
+        /* We may have just discovered more work when we were about to run out.
+           Reset waterline so that we try to refill the buffer again. */
+        pb_fill(&pb);
       }
-      budget--;
-      mark_slice_darken(stk, *me.start, &budget);
-      me.start++;
     }
-    budget--; /* credit for header */
```

Comment:
When trying to review this PR, I must say that I fail to follow the work/budget accounting logic that was already there. In the original do_some_marking function, there are two places were the budget is decremented:

https://github.com/ocaml/ocaml/blob/67e998bc95348a714c810c2b713d9f9ba180b285/runtime/major_gc.c#L736-L751

1. The second/outer decrement says that the budget update accounts for reading some block's header, but which block/header is it talking about?
2. What does the first/inner decrement correspond to?
3. Where are those decrements found in the new version of the code? I see one in `do_some_marking` after the call to `mark_slice_darken_only`, but it is unclear again which header it would correspond to.

(Why not decrement the budget exactly in the place that actually reads the header, that is in `mark_slice_darken(_only)`?)

Reply:
Yes, the budget accounting is also a tricky part I avoided to modify.
1. The outer decrement is accounting for the header of the block that was pushed onto the `mark_stack` (the header of `block` from `mark_stack_push_block`). The `mark_entry` only points to the `block`'s fields.
Actually the comment at https://github.com/ocaml/ocaml/blob/67e998bc95348a714c810c2b713d9f9ba180b285/runtime/major_gc.c#L667-L668 confirm this.
2. The inner decrement account for the the fields we are reading while going over `me.start` to m`e.end `

Reply:
3. The outer decrement is https://github.com/ocaml/ocaml/blob/9767aa8c11feef5038c44efd3673a6e52f9519ef/runtime/major_gc.c#L837
And I think I've accidentally dropped the inner decrement! Well spotted again!